corehq/apps/hqcase/api/get_list.py

```diff
@@ -0,0 +1,96 @@
+import datetime
+from itertools import chain
+
+from dateutil.parser import parse
+
+from dimagi.utils.parsing import FALSE_STRINGS
+
+from corehq.apps.es import case_search
+from corehq.apps.es import cases as case_es
+
+from .core import UserError, serialize_es_case
+
+DEFAULT_PAGE_SIZE = 20
+MAX_PAGE_SIZE = 5000
+CUSTOM_PROPERTY_PREFIX = 'property.'
+
+
+def _to_boolean(val):
+    return not (val == '' or val.lower() in FALSE_STRINGS)
+
+
+def _to_int(val, param_name):
+    try:
+        return int(val)
+    except ValueError:
+        raise UserError(f"'{val}' is not a valid value for '{param_name}'")
+
+
+def _make_date_filter(date_filter, param):
+
+    def filter_fn(val):
+        try:
+            # If it's only a date, don't turn it into a datetime
+            val = datetime.datetime.strptime(val, '%Y-%m-%d').date()
+        except ValueError:
+            try:
+                val = parse(val)
+            except ValueError:
+                raise UserError(f"Cannot parse datetime '{val}'")
+        return date_filter(**{param: val})
+
+    return filter_fn
+
+
+def _to_date_filters(field, date_filter):
+    return [
+        (f'{field}.gt', _make_date_filter(date_filter, 'gt')),
+        (f'{field}.gte', _make_date_filter(date_filter, 'gte')),
+        (f'{field}.lte', _make_date_filter(date_filter, 'lte')),
+        (f'{field}.lt', _make_date_filter(date_filter, 'lt')),
+    ]
+
+
+FILTERS = {
+    'external_id': case_search.external_id,
+    'case_type': case_es.case_type,
+    'owner_id': case_es.owner,
+    'case_name': case_es.case_name,
+    'closed': lambda val: case_es.is_closed(_to_boolean(val)),
+}
+FILTERS.update(chain(*[
+    _to_date_filters('last_modified', case_es.modified_range),
+    _to_date_filters('server_last_modified', case_es.server_modified_range),
+    _to_date_filters('date_opened', case_es.opened_range),
+    _to_date_filters('date_closed', case_es.closed_range),
+]))
+
+
+def get_list(domain, params):
```

Comment:
what do you think of making `params` a namedtuple / dataclass / `attrs` class so that it's more clear what the expectations are. You can also do more validation etc in that class:
```
params = CaseListParams.from_querydict(querydict)
params.validate()  # check page_size and filter names
params.offset
> 0
params.limit
> 1000
params.custom_filters
> {}
params.standard_filters
> {}
```

Reply:
Yeah, I'll give that a go.  I tried something like that initially, but it seemed like overkill, though now with the rest of this built out I agree it might make things clearer.

Reply:
I took a crack at this, though I'm not sure it's an improvement - I'd be interested in suggestions if you've got 'em.  You can see the change here:
https://github.com/dimagi/commcare-hq/commit/74b3d0c0803

Honestly after fiddling with this some more, I'm tempted to try out something more like the [`query_param_consumers`](https://github.com/dimagi/commcare-hq/blob/2b8a63d094f99a8282fc27ff9e877d3eba2f59e4/corehq/apps/api/es.py#L497-L497) approach used in the old API.

The thing is that each parameter is optional, a few need type coercion or validation on the values, and the date parameters correspond to multiple user-supplied parameters, and we should to identify parameters which don't exist, which incorrectly add qualifiers like `gte`, or which exclude such qualifiers.  I do like doing a single pass to validate the keys, values, and assemble them into an ES filter, though it'd be nice to have the ES part somewhat separable.

Reply:
Took some time this morning to try out a third approach: https://github.com/dimagi/commcare-hq/commit/1143c7e969602e410246eb251164dae059598949